runtime/lua/vim/_defaults.lua

```diff
@@ -925,6 +925,57 @@ do
       end
     end
   end
+
+  vim.api.nvim_create_autocmd('DiagnosticChanged', {
+    pattern = '*',
+    desc = 'Update buffer diagnostics in the statusline',
+    group = vim.api.nvim_create_augroup('nvim.statusline', {}),
+    callback = function()
+      local bufnr = vim.api.nvim_get_current_buf()
+      local diagnostics = vim.diagnostic.get(bufnr)
+
+      local counts = { errors = 0, warnings = 0, info = 0, hints = 0 }
+      local severity = vim.diagnostic.severity
+
+      for _, d in ipairs(diagnostics) do
+        if d.severity == severity.ERROR then
+          counts.errors = counts.errors + 1
+        elseif d.severity == severity.WARN then
+          counts.warnings = counts.warnings + 1
+        elseif d.severity == severity.INFO then
+          counts.info = counts.info + 1
+        elseif d.severity == severity.HINT then
+          counts.hints = counts.hints + 1
+        end
+      end
+
+      local result = {}
+      if counts.errors > 0 then
+        table.insert(result, 'e:' .. counts.errors)
+      end
+      if counts.warnings > 0 then
+        table.insert(result, 'w:' .. counts.warnings)
+      end
+      if counts.info > 0 then
+        table.insert(result, 'i:' .. counts.info)
+      end
+      if counts.hints > 0 then
+        table.insert(result, 'h:' .. counts.hints)
+      end
+
+      local result_str = table.concat(result, ' ')
+
+      if #result_str > 0 then
+        result_str = result_str .. '  '
+      end
+
+      vim.b[bufnr].buffer_diagnostics = result_str
+
+      vim.schedule(function()
+        vim.cmd.redrawstatus()
+      end)
```

Comment:
Not sure if this is a good idea to do by default as it can either lead to more redraws than necessary or in the worst case interfere with messages that are being polled (like lsp status)

I think ideally the full `DiagnosticChanged` autocmd should only be setup on demand if the default status line is actually used. Otherwise everyone pays the cost for the string formatting.

Reply:
The schedule is most likely unnecessary here. I initially added it to my custom statusline because of some plugin compatibility issue. I can test this implementation without scheduling redraws.

Is there a preferred way of detecting if statusline was changed from default? I can only think of using `OptionSet` autocommand for statusline

Reply:
> I think ideally the full `DiagnosticChanged` autocmd should only be setup on demand if the default status line is actually used. Otherwise everyone pays the cost for the string formatting.

Oh, should we introduce this as `vim.diagnostic.status()` (compare `vim.lsp.status()`) ?

And yes, making this conditional on whether the default statusline is used, makes sense.

Reply:
> Oh, should we introduce this as vim.diagnostic.status() (compare vim.lsp.status()) ?

Makes sense. Looking at the implementation `vim.lsp.status()` displays messages from all clients for all buffers. `vim.diagnostic.status()` would have to be local to the current or passed in buffer for starters, can be expanded to display all diagnostics if needed.

If we decided to go with `vim.diagnostic.status()` would it be used in the statusline as is i.e. 
```lua
vim.opt.status = '... %{luaeval("vim.diagnostic.status()")} ...'
```

> And yes, making this conditional on whether the default statusline is used, makes sense.

I've updated the code to sub/unsub when statusline is changed

Reply:
> `vim.lsp.status()` displays messages from all clients for all buffers.

vim.diagnostics.status() doesn't need to exactly match how lsp.status works; it's just a reference for the "shape" + name.



> If we decided to go with `vim.diagnostic.status()` would it be used in the statusline as is i.e.

I think so. And then the default 'statusline' will reflect this too, which is Good. If there are "performance" issues, those need to be fixed internally in `status()`, e.g. by debouncing or whatever.